### PR TITLE
[21812 ] Add new verbose modes and search topics using wildcards

### PR DESCRIPTION
**Pull request by LuciaEchevarria99**
Merge after: 
- https://github.com/eProsima/DDS-Pipe/pull/134

### CODE REVIEW

**Comment by rsanchez15** [comment]
Check CI [here](https://github.com/eProsima/Fast-DDS-spy/actions/runs/11275566432/job/31357319573)

File: fastddsspy_tool/src/cpp/tool/Controller.cpp [hunk 1]
```diff
@@ -348,8 +418,13 @@ void Controller::print_command_(
...
-        if (topic.m_topic_name != arguments[1])
+        std::string topic_name = arguments[1];
+
+        ddspipe::core::types::WildcardDdsFilterTopic filter_topic;
+        filter_topic.topic_name = topic_name;
+
+        std::set<eprosima::ddspipe::core::types::DdsTopic> topics = participants::ModelParser::get_topics(*model_, filter_topic);
+        if (topics.empty())
         {
             view_.show_error(STR_ENTRY
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/src/cpp/tool/Controller.cpp:429]: This comment might not make much sense with wildcards right?

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp [hunk 2]
```diff
@@ -42,14 +42,15 @@ bool DataStreamer::activate_all(
 }
 
 bool DataStreamer::activate(
-        const ddspipe::core::types::DdsTopic& topic_to_activate,
+        const ddspipe::core::types::WildcardDdsFilterTopic& topic_to_activate,
+        const std::set<eprosima::ddspipe::core::types::DdsTopic>& topics,
         const std::shared_ptr<CallbackType>& callback)
 {
-    if (!is_topic_type_discovered(topic_to_activate))
+    if (!is_any_topic_type_discovered(topics))
```
Comment (juanlofer-eprosima) [inline fastddsspy_participants/src/cpp/model/DataStreamer.cpp:49]: I believe it would be better to have this in the tool (Controller.cpp), it seems a bit weird to pass the topics like that.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp [hunk 3]
```diff
@@ -42,14 +42,15 @@ bool DataStreamer::activate_all(
...
     {
         EPROSIMA_LOG_WARNING(FASTDDSSPY_DATASTREAMER,
```
Comment (juanlofer-eprosima) [inline fastddsspy_participants/src/cpp/model/DataStreamer.cpp:51]: This warning also seems a bit odd for the wildcard case.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp [hunk 4]
```diff
@@ -117,12 +118,13 @@ void DataStreamer::add_data(
         }
         else
         {
-            if (!(activated_topic_ == topic))
+            if (!(activated_topic_.matches(topic)))
```
Comment (juanlofer-eprosima) [inline fastddsspy_participants/src/cpp/model/DataStreamer.cpp:121]: Please fix warning above ("but schema not is available" -> "but schema is not available")

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp [hunk 5]
```diff
@@ -117,12 +118,13 @@ void DataStreamer::add_data(
...
             {
                 // If not all activated, and this is not the activated topic skip
-                EPROSIMA_LOG_WARNING(
+                EPROSIMA_LOG_INFO(
```
Comment (juanlofer-eprosima) [inline fastddsspy_participants/src/cpp/model/DataStreamer.cpp:124]: Well done, this should have never been a warning.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp [hunk 6]
```diff
@@ -142,6 +144,12 @@ void DataStreamer::add_data(
         {
             dyn_type = it->second;
         }
+
+        auto map_it = topic_type_discovered_.find(topic.m_topic_name);
```
Comment (juanlofer-eprosima) [inline fastddsspy_participants/src/cpp/model/DataStreamer.cpp:148]: Comment for line 136: watch out, this could actually occur now right? You could print topic_* , be able to activate the streamer (since for example topic_A exists and its type is discovered) and then receive a message of topic_B whose type is not discovered.

File: fastddsspy_tool/src/cpp/tool/Controller.cpp [hunk 7]
```diff
@@ -439,24 +496,27 @@ void Controller::help_command_(
...
+            << "\tparticipants <Guid>             : verbose information related with a specific DomainParticipant.\n"
+            << "\twriters                         : DataWriters discovered in the network.\n"
+            << "\twriters verbose                 : verbose information about DataWriters discovered in the network.\n"
+            << "\twriters <Guid>                  : verbose information related with a specific DataWriter.\n"
+            << "\treader                          : DataReaders discovered in the network.\n"
+            << "\treader verbose                  : verbose information about DataReaders discovered in the network.\n"
+            << "\treader <Guid>                   : verbose information related with a specific DataReader.\n"
+            << "\ttopics                          : Topics discovered in the network in compact format.\n"
+            << "\ttopics v                        : Topics discovered in the network.\n"
+            << "\ttopics vv                       : verbose information about Topics discovered in the network.\n"
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/src/cpp/tool/Controller.cpp:513]: Idea for the future: it would be helpful to have a "types" command that would be able to print the IDL representation when the TypeObject has been received.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp [hunk 8]
```diff
@@ -142,6 +144,12 @@ void DataStreamer::add_data(
...
+        if (map_it == topic_type_discovered_.end())
+        {
+            topic_type_discovered_[topic.m_topic_name] = topic.type_name;
+        }
```
Comment (juanlofer-eprosima) [inline fastddsspy_participants/src/cpp/model/DataStreamer.cpp:152]: I don't see the purpose of this.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Implementation review.

File: fastddsspy_tool/src/cpp/tool/Controller.cpp (hunk 7 above)
Comment (rsanchez15) [reply fastddsspy_tool/src/cpp/tool/Controller.cpp:513]: Yes! I thought the same when I was reviewing the _topics_ command. See [21850](https://eprosima.easyredmine.com/issues/21850)

File: fastddsspy_participants/include/fastddsspy_participants/visualization/ModelParser.hpp [hunk 9]
```diff
@@ -61,20 +63,27 @@ struct ModelParser
             const ddspipe::core::types::Guid& guid) noexcept;
 
     FASTDDSSPY_PARTICIPANTS_DllAPI
-    static ddspipe::core::types::DdsTopic get_topic(
+    static std::set<eprosima::ddspipe::core::types::DdsTopic> get_topics(
+            const SpyModel& model,
+            const ddspipe::core::types::WildcardDdsFilterTopic& filter_topic = ddspipe::core::types::WildcardDdsFilterTopic()) noexcept;
```
[the remark this answers is not in the document]
Comment (rsanchez15) [reply fastddsspy_participants/include/fastddsspy_participants/visualization/ModelParser.hpp]: Addressed in 3b8cf55

File: docs/rst/user_manual/commands/topic.rst [hunk 10]
```diff
@@ -18,7 +18,7 @@ These are the key-words recognize as this command:
 Arguments
 =========
 
-**Topic** command support 0 or 1 argument:
+**Topic** command support from 0 to 2 arguments:
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/commands/topic.rst:21]:
```suggestion
**Topic** command supports from 0 to 2 arguments:
```

File: docs/rst/user_manual/commands/topic.rst [hunk 11]
```diff
@@ -30,8 +30,7 @@ The output format is as follows: :ref:`user_manual_command_topic_output_simple`.
 Verbose
 -------
 
-This argument queries for more complete information about each of the topics in the network.
-It adds the Guid of each endpoint on the topic and the whether the type has been discovered.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/commands/topic.rst:34]: Why was this removed? Still holds right?

File: docs/rst/user_manual/commands/topic.rst [hunk 12]
```diff
@@ -47,15 +46,28 @@ This Guid must exist inside the DDS network.
     If there are 2 topics with the same name and different Topic Data Type, only one of them could be visible.
     These is a circumstance that :term:`DDS` allows, but it is strongly suggested not to do.
 
+Topic name with wildcards
+-------------------------
+
+When a topic name is with wildcards (*), this command retrieves :ref:`verbose information <user_manual_command_topic_output_verbose>` for all topics that matches the given filter.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/commands/topic.rst:52]:
```suggestion
When a topic name contains wildcards (*), this command retrieves :ref:`verbose information <user_manual_command_topic_output_verbose>` for all topics that match the given filter.
```

File: docs/rst/user_manual/usage_example.rst [hunk 13]
```diff
@@ -118,24 +114,27 @@ Try out all the commands DDS Spy has to offer:
...
-        reader                 : DataReaders discovered in the network.
-        reader verbose         : verbose information about DataReaders discovered in the network.
-        reader <Guid>          : verbose information related with a specific DataReader.
-        topics                 : Topics discovered in the network.
-        topics verbose         : verbose information about Topics discovered in the network.
-        topics <name>          : verbose information related with a specific Topic.
-        show <name>            : data of a specific Topic (Data Type must be discovered).
-        show <name> verbose    : data with additional source info of a specific Topic.
-        show all               : verbose data of all topics (only those whose Data Type is discovered).
+        help                            : this help.
```
Comment (juanlofer-eprosima) [inline docs/rst/user_manual/usage_example.rst:117]: Update after addressing request from previous revision.

File: fastddsspy_tool/test/application/test_cases/one_shot_help.py [hunk 14]
```diff
@@ -39,26 +39,27 @@ def __init__(self):
...
-\treader verbose         : verbose information about DataReaders discovered in the network.\n\
-\treader <Guid>          : verbose information related with a specific DataReader.\n\
-\ttopics                 : Topics discovered in the network.\n\
-\ttopics verbose         : verbose information about Topics discovered in the network.\n\
-\ttopics <name>          : verbose information related with a specific Topic.\n\
-\tshow <name>            : data of a specific Topic (Data Type must be discovered).\n\
-\tshow <name> verbose    : data with additional source info of a specific Topic.\n\
-\tshow all               : verbose data of all topics (only those whose Data Type is \
-discovered).\n\
+\thelp                            : this help.\n\
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/test/application/test_cases/one_shot_help.py:42]: Update.

File: fastddsspy_tool/test/application/test_cases/one_shot_help_dds.py [hunk 15]
```diff
@@ -42,26 +42,27 @@ def __init__(self):
...
-\treader verbose         : verbose information about DataReaders discovered in the network.\n\
-\treader <Guid>          : verbose information related with a specific DataReader.\n\
-\ttopics                 : Topics discovered in the network.\n\
-\ttopics verbose         : verbose information about Topics discovered in the network.\n\
-\ttopics <name>          : verbose information related with a specific Topic.\n\
-\tshow <name>            : data of a specific Topic (Data Type must be discovered).\n\
-\tshow <name> verbose    : data with additional source info of a specific Topic.\n\
-\tshow all               : verbose data of all topics (only those whose Data Type is \
-discovered).\n\
+\thelp                            : this help.\n\
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/test/application/test_cases/one_shot_help_dds.py:45]: Update.

File: fastddsspy_tool/test/application/test_cases/tool_help.py [hunk 16]
```diff
@@ -37,37 +37,36 @@ def __init__(self):
             arguments_dds=[],
             arguments_spy=[],
             commands_spy=['help'],
-            output=""">> \x1b[0mFast DDS Spy is an interactive CLI that allow to instrospect\
- DDS networks.\n\n\
+            output=""">> \x1b[0mFast DDS Spy is an interactive CLI that allow to instrospect DDS networks.\n\n\
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/test/application/test_cases/tool_help.py:40]: Update.

File: fastddsspy_tool/test/application/test_cases/tool_help_dds.py [hunk 17]
```diff
@@ -40,37 +40,36 @@ def __init__(self):
             arguments_dds=[],
             arguments_spy=['--config-path', 'configuration'],
             commands_spy=['help'],
-            output=""">> \x1b[0mFast DDS Spy is an interactive CLI that allow to instrospect\
- DDS networks.\n\n\
+            output=""">> \x1b[0mFast DDS Spy is an interactive CLI that allow to instrospect DDS networks.\n\n\
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/test/application/test_cases/tool_help_dds.py:43]: Update.

File: fastddsspy_tool/test/application/test_class.py [hunk 18]
```diff
@@ -208,6 +209,7 @@ def valid_output(self, output) -> bool:
         lines_expected_output = expected_output.splitlines()
         lines_output = output.splitlines()
 
+        # TODO (Raul): If guid and rate are on the same line this will not work.
```
Comment (juanlofer-eprosima) [inline fastddsspy_tool/test/application/test_class.py:212]: Address?

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Tests and documentation revision.

File: fastddsspy_tool/src/cpp/tool/Controller.cpp (hunk 1 above)
Comment (rsanchez15) [reply fastddsspy_tool/src/cpp/tool/Controller.cpp:432]: Updated

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp (hunk 2 above)
Comment (rsanchez15) [reply fastddsspy_participants/src/cpp/model/DataStreamer.cpp:49]: I do not understand your comment about this

Comment (juanlofer-eprosima) [reply fastddsspy_participants/src/cpp/model/DataStreamer.cpp:49]: This check used to be done both in the controller (Controller::print_command_), and then again here. I asked Lucía to keep it only in one place, and she left it here, but I feel it would be better only in the controller. Two reasons:
- The behaviour it's the same but the way it is now is more restrictive, and I think it's better to let the user of this class (controller in this case) to set the restriction.
- To avoid having to pass topics as argument, which seems a bit odd to me.

File: docs/rst/user_manual/commands/topic.rst (hunk 11 above)
Comment (rsanchez15) [reply docs/rst/user_manual/commands/topic.rst:34]: The guid is only for the higher verbosity mode

File: fastddsspy_tool/test/application/test_class.py (hunk 18 above)
Comment (rsanchez15) [reply fastddsspy_tool/test/application/test_class.py:212]: This should be addressed in another PR, as I detected it here, but it has remained this way since the test implementation.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp (hunk 2 above)
Comment (juanlofer-eprosima) [reply fastddsspy_participants/src/cpp/model/DataStreamer.cpp:49]: Not yet applied.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp (hunk 6 above)
Comment (juanlofer-eprosima) [reply fastddsspy_participants/src/cpp/model/DataStreamer.cpp:150]: Better know with a warning instead, but please also remove the code comment.

File: fastddsspy_tool/src/cpp/tool/Controller.cpp [hunk 19]
```diff
@@ -282,30 +293,89 @@ void Controller::topics_command_(
...
+                                << "<"
+                                << arguments[1]
+                                << "> does not match any topic in the DDS network.");
+                        return;
+                    }
+
+                    ddspipe::yaml::set_collection(yml, data);
+                }
+            }
+            else
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply fastddsspy_tool/src/cpp/tool/Controller.cpp:372]: Ok, but I would prefer the condition is made explicit. Maybe you could even move the == 2 case to be before the == 3 one.

File: fastddsspy_participants/src/cpp/model/DataStreamer.cpp (hunk 2 above)
Comment (LuciaEchevarria99) [reply fastddsspy_participants/src/cpp/model/DataStreamer.cpp:49]: I agree

File: docs/rst/user_manual/commands/data.rst [hunk 20]
```diff
@@ -30,25 +30,36 @@ The information whether the topic data type is already discovered be the applica
...
+For example, if you enter ``sensor_*``, the command will print data from all topics starting with ``sensor_``, such as ``sensor_temperature``, ``sensor_humidity``, etc.
 
 Topic name + Verbose
 --------------------
 
 Giving a topic name and the the :ref:`verbose argument <user_manual_commands_input_verbose>` the output is the data received in real-time with additional meta-information as the topic name, the source timestamp, and the source :term:`DataWriter` :term:`Guid`.
-Data is printing using :ref:`user_manual_command_show_output_verbose`.
+Data is printing using :ref:`user_manual_command_echo_output_verbose`.
+
+Topic name prefix + * + Verbose
```
[the remark this answers is not in the document]
Comment (LuciaEchevarria99) [reply docs/rst/user_manual/commands/data.rst:53]: Yes, you are right, I will add `* + Topic name sufix + Verbose` case to the title

File: docs/rst/user_manual/commands/topic.rst (hunk 11 above)
Comment (juanlofer-eprosima) [reply docs/rst/user_manual/commands/topic.rst:34]: So we are only documenting what information is provided through examples? Maybe it should be explained what extra information is showed in verbose mode, and in super verbose mode.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: docs/rst/user_manual/commands/data.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (rsanchez15) [reply docs/rst/user_manual/commands/data.rst]: That's not the point. The wildcard will apply also if you have cases like ``to*ic``.

[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/user_manual/commands/data.rst]: Please update also helper (and everywhere it appears).

File: docs/rst/user_manual/commands/data.rst [hunk 21]
```diff
@@ -30,25 +30,36 @@ The information whether the topic data type is already discovered be the applica
...
+For example, if you enter ``sensor_*``, the command will print data from all topics starting with ``sensor_``, such as ``sensor_temperature``, ``sensor_humidity``, etc.
 
 Topic name + Verbose
 --------------------
 
 Giving a topic name and the the :ref:`verbose argument <user_manual_commands_input_verbose>` the output is the data received in real-time with additional meta-information as the topic name, the source timestamp, and the source :term:`DataWriter` :term:`Guid`.
-Data is printing using :ref:`user_manual_command_show_output_verbose`.
+Data is printing using :ref:`user_manual_command_echo_output_verbose`.
+
+Topic name prefix + * + Verbose / * + Topic name sufix + Verbose
```
[the remark this answers is not in the document]
Comment (juanlofer-eprosima) [reply docs/rst/user_manual/commands/data.rst]: Working on it, same situation in other repos.

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM
